File: client/src/main/java/com/alphasystem/game/uno/client/ui/control/delegate/PlayersView.java
package com.alphasystem.game.uno.client.ui.control.delegate;

import com.alphasystem.game.uno.model.PlayerDetail;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.control.Control;
import javafx.scene.control.Skin;

import java.util.List;

public final class PlayersView extends Control {

    private final ObservableList<PlayerDetail> playerDetails = FXCollections.observableArrayList();
    private final ObjectProperty<PlayerDetail> myPlayer = new SimpleObjectProperty<>(this, "myPlayer", null);
    private final IntegerProperty selectedPosition = new SimpleIntegerProperty(this, "selectedPosition", -1);

    public PlayersView() {
        setSkin(createDefaultSkin());
    }

    @Override
    protected Skin<?> createDefaultSkin() {
        return new PlayersSkin(this);
    }

    public ObservableList<PlayerDetail> getPlayerDetails() {
        return playerDetails;
    }

    public void setPlayerDetails(List<PlayerDetail> playerDetails) {
        playerDetails.forEach(playerDetail -> getPlayerDetails().add(playerDetail));
    }

    public void addPlayer(PlayerDetail playerDetail) {
        playerDetails.add(playerDetail.position(), playerDetail);
    }

    public PlayerDetail getMyPlayer() {
        return myPlayer.get();
    }

    public ObjectProperty<PlayerDetail> myPlayerProperty() {
        return myPlayer;
    }

    public void setMyPlayer(PlayerDetail myPlayer) {
        this.myPlayer.set(myPlayer);
    }

    public int getSelectedPosition() {
        return selectedPosition.get();
    }

    public IntegerProperty selectedPositionProperty() {
        return selectedPosition;
    }

    public void setSelectedPosition(int selectedPosition) {
        this.selectedPosition.set(selectedPosition);
    }
}

File: client/src/main/scala/com/alphasystem/game/uno/client/ui/control/PlayersView.scala
package com.alphasystem.game.uno.client.ui.control

import com.alphasystem.game.uno.client.ui.control.delegate.{PlayersView => JPlayersView}
import com.alphasystem.game.uno.model.PlayerDetail
import javafx.collections.ObservableList
import scalafx.Includes._
import scalafx.beans.property._
import scalafx.scene.Node

import scala.jdk.CollectionConverters._

class PlayersView private(override val delegate: JPlayersView) extends Node(delegate) {

  def numberOfPlayers: Int = playerDetails.size()

  def playerDetails: ObservableList[PlayerDetail] = delegate.getPlayerDetails

  def playerDetails_=(playerDetails: List[PlayerDetail]): Unit = delegate.setPlayerDetails(playerDetails.asJava)

  def addPlayer(playerDetail: PlayerDetail): Unit = delegate.addPlayer(playerDetail)

  def removePlayer(playerDetail: PlayerDetail): Unit = delegate.removePlayer(playerDetail)

  def myPlayer: PlayerDetail = delegate.getMyPlayer

  def myPlayer_=(playerDetail: PlayerDetail): Unit = delegate.setMyPlayer(playerDetail)

  def selectedPlayer: ObjectProperty[PlayerDetail] = delegate.selectedPlayerProperty()

  def selectedPlayer_=(playerDetail: PlayerDetail): Unit = delegate.setSelectedPlayer(playerDetail)
}

object PlayersView {
  def apply(): PlayersView = new PlayersView(new JPlayersView())
}
